### PR TITLE
CSHARP-4293: Update with aggregation pipeline fails when used on a OfTypeMongoCollection (OfType<T>).

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

File: src/MongoDB.Driver/FilteredMongoCollectionBase.cs [hunk 1]
```diff
@@ -351,60 +351,62 @@ protected virtual UpdateDefinition<TDocument> AdjustUpdateDefinition(UpdateDefin
...
+                                ArrayFilters = updateOneModel.ArrayFilters,
+                                Collation = updateOneModel.Collation,
+                                Hint = updateOneModel.Hint,
+                                IsUpsert = updateOneModel.IsUpsert
+                            };
+                        default:
+                            throw new MongoInternalException("Request type is invalid.");
+                    }
+                })
+                .ToList();
```
Comment (JamesKovacs) [inline src/MongoDB.Driver/FilteredMongoCollectionBase.cs:409]: It appears that the only change is the addition of `ToList()` on the end. This makes the diff confusing to review.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Just need to fix the diff as it looks like more was changed than actually was due to reformatting.

File: src/MongoDB.Driver/FilteredMongoCollectionBase.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/FilteredMongoCollectionBase.cs]: done

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Comment by DmitryLukyanov** [comment]
This provided solution is incorrect, instead converting a pipeline argument to regular update document, we should emulate `$SetOnInsert` with aggregate pipeline stage. This was recommended by query team:

    rs:PRIMARY> db.mycoll.update({}, [{$set: {x: {$cond: [{$eq: [{$type: "$_id"}, "missing"]}, 1, 2]}}}], {upsert: true})
    WriteResult({
            "nMatched" : 0,
            "nUpserted" : 1,
            "nModified" : 0,
            "_id" : ObjectId("630561dceb783de551833de9")
    })
    rs:PRIMARY> db.mycoll.find()
    { "_id" : ObjectId("630561dceb783de551833de9"), "x" : 1 }
    rs:PRIMARY> db.mycoll.update({}, [{$set: {x: {$cond: [{$eq: [{$type: "$_id"}, "missing"]}, 1, 2]}}}], {upsert: true})
    WriteResult({ "nMatched" : 1, "nUpserted" : 0, "nModified" : 1 })
    rs:PRIMARY> db.mycoll.find()
    { "_id" : ObjectId("630561dceb783de551833de9"), "x" : 2 }

**Comment by DmitryLukyanov** [comment]
ready for review

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Overall looking good, but we have failures on 3.6 and 4.0. From the EG logs:
```
FAILURE: MongoDB.Driver.MongoCommandException : Command update failed: BSON field 'update.updates.u' is the wrong type 'array', expected type 'object'.
```

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**